stdlib/universe/tail.go

```diff
@@ -0,0 +1,244 @@
+	flux.RegisterOpSpec(TailKind, newTailOp)
+	plan.RegisterProcedureSpec(TailKind, newTailProcedure, TailKind)
+	execute.RegisterTransformation(TailKind, createTailTransformation)
+}
+
+func createTailOpSpec(args flux.Arguments, a *flux.Administration) (flux.OperationSpec, error) {
+	if err := a.AddParentFromArgs(args); err != nil {
+		return nil, err
+	}
+
+	spec := new(TailOpSpec)
+
+	n, err := args.GetRequiredInt("n")
+	if err != nil {
+		return nil, err
+	}
+	spec.N = n
+
+	if offset, ok, err := args.GetInt("offset"); err != nil {
+		return nil, err
+	} else if ok {
+		spec.Offset = offset
+	}
+
+	return spec, nil
+}
+
+func newTailOp() flux.OperationSpec {
+	return new(TailOpSpec)
+}
+
+func (s *TailOpSpec) Kind() flux.OperationKind {
+	return TailKind
+}
+
+type TailProcedureSpec struct {
+	plan.DefaultCost
+	N      int64 `json:"n"`
+	Offset int64 `json:"offset"`
+}
+
+func newTailProcedure(qs flux.OperationSpec, pa plan.Administration) (plan.ProcedureSpec, error) {
+	spec, ok := qs.(*TailOpSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", qs)
+	}
+	return &TailProcedureSpec{
+		N:      spec.N,
+		Offset: spec.Offset,
+	}, nil
+}
+
+func (s *TailProcedureSpec) Kind() plan.ProcedureKind {
+	return TailKind
+}
+func (s *TailProcedureSpec) Copy() plan.ProcedureSpec {
+	ns := new(TailProcedureSpec)
+	*ns = *s
+	return ns
+}
+
+// TriggerSpec implements plan.TriggerAwareProcedureSpec
+func (s *TailProcedureSpec) TriggerSpec() plan.TriggerSpec {
+	return plan.NarrowTransformationTriggerSpec{}
+}
+
+func createTailTransformation(id execute.DatasetID, mode execute.AccumulationMode, spec plan.ProcedureSpec, a execute.Administration) (execute.Transformation, execute.Dataset, error) {
+	s, ok := spec.(*TailProcedureSpec)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid spec type %T", spec)
+	}
+	cache := execute.NewTableBuilderCache(a.Allocator())
+	d := execute.NewDataset(id, mode, cache)
+	t := NewTailTransformation(d, cache, s)
+	return t, d, nil
+}
+
+type tailTransformation struct {
+	d     execute.Dataset
+	cache execute.TableBuilderCache
+
+	n, offset int
+}
+
+func NewTailTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *TailProcedureSpec) *tailTransformation {
+	return &tailTransformation{
+		d:      d,
+		cache:  cache,
+		n:      int(spec.N),
+		offset: int(spec.Offset),
+	}
+}
+
+func (t *tailTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *tailTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("tail found duplicate table with key: %v", tbl.Key())
+	}
+	if err := execute.AddTableCols(tbl, builder); err != nil {
+		return err
+	}
+
+	cols := tbl.Cols()
+	isEmpty := false
+
+	n := t.n
+	offset := t.offset
+	index := 0
+	window := make([][]values.Value, n+offset)
```

Comment:
Instead of allocating a table of values, would it be possible to calculate the beginning and ending row indices, and then use [appendSlicedCols](https://github.com/influxdata/flux/blob/11215b685510974f3742a649449ec5a1bf9d03e9/stdlib/universe/limit.go#L177-L232)? This way we don't have to allocate any untracked memory.